File: examples/Microsoft.Spark.FSharp.Examples/Sql/Streaming/StructuredNetworkWordCountWindowed.fs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

namespace Microsoft.Spark.Examples.Sql.Streaming

open Microsoft.Spark.Examples
open Microsoft.Spark.Sql

/// <summary>
/// The example is taken/modified from
/// spark/examples/src/main/python/sql/streaming/structured_network_wordcount_windowed.py
///
/// You can set up the data source as follow in a separated terminal:
/// `$ nc -lk 9999`
/// to start writing standard input to port 9999.
/// </summary>
type StructuredNetworkWordCountWindowed() =
    member this.Run(args : string[]) =
        match args with
        | ([| hostname; portStr; windowSizeStr |] | [| hostname; portStr; windowSizeStr; _ |]) ->
            let port = portStr |> int64
            let windowSize = windowSizeStr |> int64
            let slideSize = if (args.Length = 3) then windowSize else (args.[3] |> int64)
            if (slideSize > windowSize) then
                printfn "<slide duration> must be less than or equal to <window duration>"
            let windowDuration = sprintf "%d seconds" windowSize
            let slideDuration = sprintf "%d seconds" slideSize

            let spark =
                SparkSession.Builder().AppName("StructuredNetworkWordCountWindowed").GetOrCreate()

            let lines =
                spark.ReadStream()
                    .Format("socket")
                    .Option("host", hostname)
                    .Option("port", port)
                    .Option("includeTimestamp", true)
                    .Load()

            let words =
                lines.Select(Functions.Explode(Functions.Split(lines.["value"], " "))
                    .Alias("word"), lines.["timestamp"])
            let windowedCounts =
                words.GroupBy(Functions.Window(words.["timestamp"], windowDuration, slideDuration),
                        words.["word"])
                    .Count()
                    .OrderBy("window")

            let query =
                windowedCounts.WriteStream()
                    .OutputMode("complete")
                    .Format("console")
                    .Option("truncate", false)
                    .Start()

            query.AwaitTermination()
            0
        | _ ->
            printfn "Usage: StructuredNetworkWordCountWindowed \
                     <hostname> <port> <window duration in seconds> \
                     [<slide duration in seconds>]"
            1

    interface IExample with
        member this.Run (args) = this.Run (args)

File: examples/Microsoft.Spark.CSharp.Examples/Sql/Streaming/StructuredNetworkWordCountWindowed.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using Microsoft.Spark.Sql;
using static Microsoft.Spark.Sql.Functions;

namespace Microsoft.Spark.Examples.Sql.Streaming
{
    /// <summary>
    /// The example is taken/modified from
    /// spark/examples/src/main/python/sql/streaming/structured_network_wordcount_windowed.py
    ///
    /// You can set up the data source as follow in a separated terminal:
    /// `$ nc -lk 9999`
    /// to start writing standard input to port 9999.
    /// </summary>
    internal sealed class StructuredNetworkWordCountWindowed : IExample
    {
        public void Run(string[] args)
        {
            if (args.Length != 3 && args.Length != 4)
            {
                Console.Error.WriteLine(
                    "Usage: StructuredNetworkWordCountWindowed " +
                    "<hostname> <port> <window duration in seconds> " +
                    "[<slide duration in seconds>]");
                Environment.Exit(1);
            }

            string hostname = args[0];
            int port = int.Parse(args[1]);
            int windowSize = int.Parse(args[2]);
            int slideSize = (args.Length == 3) ? windowSize : int.Parse(args[3]);
            if (slideSize > windowSize)
            {
                Console.Error.WriteLine(
                    "<slide duration> must be less than or equal " +
                    "to <window duration>");
            }
            string windowDuration = $"{windowSize} seconds";
            string slideDuration = $"{slideSize} seconds";

            SparkSession spark = SparkSession
                .Builder()
                .AppName("StructuredNetworkWordCountWindowed")
                .GetOrCreate();

            DataFrame lines = spark
                .ReadStream()
                .Format("socket")
                .Option("host", hostname)
                .Option("port", port)
                .Option("includeTimestamp", true)
                .Load();

            DataFrame words = lines
                .Select(Explode(Split(lines["value"], " "))
                    .Alias("word"), lines["timestamp"]);
            DataFrame windowedCounts = words
                .GroupBy(Window(words["timestamp"], windowDuration, slideDuration),
                    words["word"])
                .Count()
                .OrderBy("window");

            Spark.Sql.Streaming.StreamingQuery query = windowedCounts
                .WriteStream()
                .OutputMode("complete")
                .Format("console")
                .Option("truncate", false)
                .Start();

            query.AwaitTermination();
        }
    }
}
